relay-profiling/src/sample.rs

```diff
@@ -228,27 +228,27 @@ impl SampleProfile {
     }
 
     fn remove_idle_samples_at_the_edge(&mut self) {
-        self.remove_idle_samples_at_the_start();
-        self.profile.samples.reverse();
-        self.remove_idle_samples_at_the_start();
-        self.profile.samples.reverse();
-    }
+        if let Some(start) = self.profile.samples.iter().position(|sample| {
+            match self.profile.stacks.get(sample.stack_id) {
+                Some(stack) => !stack.is_empty(),
+                None => false,
+            }
+        }) {
+            if start > 0 {
+                self.profile.samples.drain(..start);
+            }
+        }
 
-    fn remove_idle_samples_at_the_start(&mut self) {
-        let mut thread_ids: HashSet<u64> = HashSet::new();
-        self.profile
-            .samples
-            .retain(|sample| match self.profile.stacks.get(sample.stack_id) {
-                Some(stack) => {
-                    if stack.is_empty() && !thread_ids.contains(&sample.thread_id) {
-                        false
-                    } else {
-                        thread_ids.insert(sample.thread_id);
-                        true
-                    }
-                }
-                None => true,
-            });
+        if let Some(end) = self.profile.samples.iter().rev().position(|sample| {
```

Comment:
As far as I know, `.rev().position()` gives you the offset from the end, so I think we need `.rposition()` here.

```suggestion
        if let Some(end) = self.profile.samples.iter().rposition(|sample| {
```

Example:

```rust
let xs = vec!['a', 'b', 'c', 'd'];
println!("{:?}", xs.iter().rev().position(|item| *item == 'c'));
println!("{:?}", xs.iter().rposition(|item| *item == 'c'));

// prints
Some(1)
Some(2)
```

You can verify this by removing the last sample from your test case -- then `end` will be `2`.